### DISCUSSION TITLE
官网国际化文档中配置key的方式(用object类型定义)，在实际使用中并不生效，是需要单独配置formatjs的解析规则吗？请问应该怎么做呢？

### DISCUSSION THREAD

**Discussion by leila-huang** [umijs/umi#11175] [Q&A]
文档中提到这种配置国际化key的方式，在实际使用中并不生效，是需要单独配置formatjs的解析规则吗？请问应该怎么做呢？
```
// src/locales/zh-CN.ts
export default {
  user: {
    welcome: '{name}，今天也是美好的一天！',
  },
};
...
  <FormattedMessage id="user.welcome" values={{ name: '张三' }} />
```

**Comment by fz6m** [comment]
`locales` 的文案需要打平，参考 ant pro ：https://github.com/ant-design/ant-design-pro/blob/master/src/locales/zh-CN/pages.ts

**Reply by leila-huang** [reply]
噢噢，那官网这个例子确实误导了...，因为目前在业务中发现‘key.key’:'value’ 的方式会蛮占空间的，把key转成对象或者缩短id都会减少一些 locales 的体积。所以也想问问为什么不用自动生成ID的方式呢，是遇到什么弊端了嘛？我看formatjs目前也是可以生成的。

**Reply by AlbertXiaoPeng** [reply]
这个官网文档也太误导了
![image](https://github.com/umijs/umi/assets/44151574/71dabee8-60c5-46eb-bf5c-b80ab2b1d2f8)
https://umijs.org/docs/max/i18n#%E5%A4%9A%E8%AF%AD%E8%A8%80%E9%BB%98%E8%AE%A4%E5%80%BC
整的我多层这种国际化出不来，要不就改下官网文档，明确说明下
> `locales` 的文案需要打平

**Reply by leila-huang** [reply]
我现在直接用自动生成ID的方式去管理了，不过源码里禁止用动态ID，这里要改一下代码，就能用了，不过我项目里面还没接入国际化，用的还是中文。不知道会不会有别的问题。之前测试中英文切换是没问题了![image](https://github.com/umijs/umi/assets/22982196/1306afa5-ce23-4f6f-b377-69d5cde709db)
![image](https://github.com/umijs/umi/assets/22982196/46f26410-a735-490e-b40c-62940f5e5925)

**Reply by fz6m** [reply]
关于文案里嵌套的场景已经有 pr 支持了，目前还未合入 #11218 。

在合并发布前，你可以通过这个代码片段打平所有文案的导出先解决：

```ts
const flattenMessages=(
  nestedMessages: Record<string, any>,
  prefix = '',
) => {
  return Object.keys(nestedMessages).reduce(
    (messages: Record<string, any>, key) => {
      const value = nestedMessages[key];
      const prefixedKey = prefix ? `${prefix}.${key}` : key;
      if (typeof value === 'string') {
        messages[prefixedKey] = value;
      } else {
        Object.assign(messages, flattenMessages(value, prefixedKey));
      }
      return messages;
    },
    {},
  );
}
```
